crates/astria-core/src/protocol/transaction/v1/action/mod.rs

```diff
@@ -907,67 +912,144 @@ pub struct Ics20Withdrawal {
     pub fee_asset: asset::Denom,
     // a memo to include with the transfer
     pub memo: String,
+    // whether to use a bech32-compatible format of the `.return_address` when generating
+    // fungible token packets (as opposed to Astria-native bech32m addresses). This is
+    // necessary for chains like noble which enforce a strict bech32 format.
+    pub use_compat_address: bool,
+}
+
+/// An IBC Withdrawal from a rollup, containing a bridge address. The sender of this kind of action
+/// must be the bridge account's withdrawer.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Ics20WithdrawalWithBridgeAddress {
+    // a transparent value consisting of an amount and a denom.
+    pub amount: u128,
+    // the asset to transfer.
+    pub denom: Denom,
+    // the address on the destination chain to send the transfer to.
+    pub destination_chain_address: String,
+    // an Astria address to use to return funds from this withdrawal
+    // in the case it fails.
+    pub return_address: Address,
+    // the height (on the counterparty chain) at which this transfer expires.
+    pub timeout_height: IbcHeight,
+    // the unix timestamp (in nanoseconds) at which this transfer expires.
+    pub timeout_time: u64,
+    // the source channel used for the withdrawal.
+    pub source_channel: ChannelId,
+    // the asset to use for fee payment.
+    pub fee_asset: asset::Denom,
+    // the parsed memo, which contains rollup information for ics20 withdrawals which contain a
+    // bridge address.
+    pub ics20_withdrawal_from_rollup: Ics20WithdrawalFromRollup,
     // the address of the bridge account to transfer from, if this is a withdrawal
     // from a bridge account and the sender of the tx is the bridge's withdrawer,
     // which differs from the bridge account's address.
-    //
-    // if unset, and the transaction sender is not a bridge account, the withdrawal
-    // is treated as a user (non-bridge) withdrawal.
-    //
-    // if unset, and the transaction sender is a bridge account, the withdrawal is
-    // treated as a bridge withdrawal (ie. the bridge account's withdrawer address is checked).
-    pub bridge_address: Option<Address>,
-
+    pub bridge_address: Address,
     // whether to use a bech32-compatible format of the `.return_address` when generating
     // fungible token packets (as opposed to Astria-native bech32m addresses). This is
     // necessary for chains like noble which enforce a strict bech32 format.
     pub use_compat_address: bool,
+    // the memo included from the raw action. This is kept so that the action can be infallibly
+    // converted back into its raw format.
```

Comment:
why is infallible conversion needed?

Reply:
All domain types (that have a wire counterpart) must be convertible to their wire format.

This is a design decision of the crate and something that would be a pain to lose. It's not yet enforced but should be through a trait impl (`Protobuf` at the moment, but we should rename that).

Reply:
like can we change `self.memo()` to return the json-serialized `ics20_withdrawal_from_rollup` if it's the `Self::FromRollup` variant?

Reply:
It's been a minute since I made these changes, so I may be misremembering, but I believe the infallibility is needed to satisfy the `Protobuf` trait implementation, and we need to store the original memo because re-serializing `ics20_withdrawal_from_rollup` is not an infallible operation.